golang/cosmos/x/lien/lien.go

```diff
@@ -0,0 +1,140 @@
+// Package lien is a cosmos-sdk module that implements liens.
+// Liens are an encumbrance that prevents the transfer of tokens
+// out of an account, much like the unvested tokens in a vesting
+// account.  See spec/ for full details.
+package lien
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Agoric/agoric-sdk/golang/cosmos/vm"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// portHandler implements a vm.PortHandler.
+type portHandler struct {
+	keeper Keeper
+}
+
+// portMessage is a struct that any lien bridge message can unmarshal into.
+type portMessage struct {
+	Type    string `json:"type"`
+	Address string `json:"address"`
+	Denom   string `json:"denom"`
+	Amount  string `json:"amount"`
+}
+
+// msgAccountState marshals into the AccountState message for the lien bridge.
+type msgAccountState struct {
+	CurrentTime string `json:"currentTime"`
+	Total       string `json:"total"`
+	Bonded      string `json:"bonded"`
+	Unbonding   string `json:"unbonding"`
+	Locked      string `json:"locked"`
+	Liened      string `json:"liened"`
+}
+
+// NewPortHandler returns a port handler for the Keeper.
+func NewPortHandler(k Keeper) vm.PortHandler {
+	return portHandler{keeper: k}
+}
+
+const (
+	LIEN_GET_ACCOUNT_STATE = "LIEN_GET_ACCOUNT_STATE"
+	LIEN_SET_TOTAL         = "LIEN_SET_TOTAL"
+)
+
+// Receive implements the vm.PortHandler method.
+// Receives and processes a bridge message, returning the
+// JSON-encoded response or error.
+// See spec/02_messages.md for the messages and responses.
+func (ch portHandler) Receive(ctx *vm.ControllerContext, str string) (string, error) {
+	var msg portMessage
+	err := json.Unmarshal([]byte(str), &msg)
+	if err != nil {
+		return "", err
+	}
+	switch msg.Type {
+	case LIEN_GET_ACCOUNT_STATE:
+		return ch.handleGetAccountState(ctx.Context, msg)
+
+	case LIEN_SET_TOTAL:
+		return ch.handleSetTotal(ctx.Context, msg)
+	}
+	return "", fmt.Errorf("unrecognized type %s", msg.Type)
+}
+
+// handleGetAccountState processes a LIEN_GET_ACCOUNT_STATE message.
+// See spec/02_messages.md for the messages and responses.
+func (ch portHandler) handleGetAccountState(ctx sdk.Context, msg portMessage) (string, error) {
+	addr, err := sdk.AccAddressFromBech32(msg.Address)
+	if err != nil {
+		return "", fmt.Errorf("cannot convert %s to address: %w", msg.Address, err)
+	}
+	denom := msg.Denom
+	if err = sdk.ValidateDenom(denom); err != nil {
+		return "", fmt.Errorf("invalid denom %s: %w", denom, err)
+	}
+	state := ch.keeper.GetAccountState(ctx, addr)
+	reply := msgAccountState{
+		CurrentTime: ctx.BlockTime().String(), // REVIEWER: what format?
```

Comment:
I believe `ctx.BlockTime()` is an int64 of seconds since the epoch.

Reply:
It's a `time.Time`, but we'll represent it as an `int64` of seconds since UNIX epoch.